File: ast2csharp/ast.cs
using System;
using System.CodeDom.Compiler;
using System.Collections.Generic;
namespace ast2cs
{
	public interface Node
	{
		public Loc Loc { get; set; }
	}
	public interface Expr : Node
	{
		public Type? ExprType { get; set; }
	}
	public interface Literal : Expr
	{
	}
	public interface Stmt : Node
	{
	}
	public interface Member : Stmt
	{
		public Member? Belong { get; set; }
		public Ident? Ident { get; set; }
	}
	public interface Type : Node
	{
		public bool IsExplicit { get; set; }
	}
	public class Program : Node
	{
		public Loc Loc { get; set; }
		public StructType? StructType { get; set; }
		public List<Node>? Elements { get; set; }
		public Scope? GlobalScope { get; set; }
	}
	public class Binary : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public BinaryOp Op { get; set; }
		public Expr? Left { get; set; }
		public Expr? Right { get; set; }
	}
	public class Unary : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public UnaryOp Op { get; set; }
		public Expr? Expr { get; set; }
	}
	public class Cond : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public Expr? Cond_ { get; set; }
		public Expr? Then { get; set; }
		public Loc ElsLoc { get; set; }
		public Expr? Els { get; set; }
	}
	public class Ident : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public string Ident_ { get; set; }
		public IdentUsage Usage { get; set; }
		public Node? Base { get; set; }
		public Scope? Scope { get; set; }
	}
	public class Call : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public Expr? Callee { get; set; }
		public Expr? RawArguments { get; set; }
		public List<Expr>? Arguments { get; set; }
		public Loc EndLoc { get; set; }
	}
	public class If : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public Scope? CondScope { get; set; }
		public Expr? Cond { get; set; }
		public IndentBlock? Then { get; set; }
		public Node? Els { get; set; }
	}
	public class MemberAccess : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public Expr? Target { get; set; }
		public string Member { get; set; }
		public Loc MemberLoc { get; set; }
		public Node? Base { get; set; }
	}
	public class Paren : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public Expr? Expr { get; set; }
		public Loc EndLoc { get; set; }
	}
	public class Index : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public Expr? Expr { get; set; }
		public Expr? Index_ { get; set; }
		public Loc EndLoc { get; set; }
	}
	public class Match : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public Scope? CondScope { get; set; }
		public Expr? Cond { get; set; }
		public List<Node>? Branch { get; set; }
	}
	public class Range : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public BinaryOp Op { get; set; }
		public Expr? Start { get; set; }
		public Expr? End { get; set; }
	}
	public class TmpVar : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public ulong TmpVar_ { get; set; }
	}
	public class BlockExpr : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public List<Node>? Calls { get; set; }
		public Expr? Expr { get; set; }
	}
	public class Import : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public string Path { get; set; }
		public Call? Base { get; set; }
		public Program? ImportDesc { get; set; }
	}
	public class IntLiteral : Literal
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public string Value { get; set; }
	}
	public class BoolLiteral : Literal
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public bool Value { get; set; }
	}
	public class StrLiteral : Literal
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public string Value { get; set; }
	}
	public class Input : Literal
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
	}
	public class Output : Literal
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
	}
	public class Config : Literal
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
	}
	public class Loop : Stmt
	{
		public Loc Loc { get; set; }
		public Scope? CondScope { get; set; }
		public Expr? Init { get; set; }
		public Expr? Cond { get; set; }
		public Expr? Step { get; set; }
		public IndentBlock? Body { get; set; }
	}
	public class IndentBlock : Stmt
	{
		public Loc Loc { get; set; }
		public List<Node>? Elements { get; set; }
		public Scope? Scope { get; set; }
		public StructType? StructType { get; set; }
	}
	public class MatchBranch : Stmt
	{
		public Loc Loc { get; set; }
		public Expr? Cond { get; set; }
		public Loc SymLoc { get; set; }
		public Node? Then { get; set; }
	}
	public class Return : Stmt
	{
		public Loc Loc { get; set; }
		public Expr? Expr { get; set; }
	}
	public class Break : Stmt
	{
		public Loc Loc { get; set; }
	}
	public class Continue : Stmt
	{
		public Loc Loc { get; set; }
	}
	public class Assert : Stmt
	{
		public Loc Loc { get; set; }
		public Binary? Cond { get; set; }
	}
	public class ImplicitYield : Stmt
	{
		public Loc Loc { get; set; }
		public Expr? Expr { get; set; }
	}
	public class Field : Member
	{
		public Loc Loc { get; set; }
		public Member? Belong { get; set; }
		public Ident? Ident { get; set; }
		public Loc ColonLoc { get; set; }
		public Type? FieldType { get; set; }
		public Expr? RawArguments { get; set; }
		public List<Expr>? Arguments { get; set; }
	}
	public class Format : Member
	{
		public Loc Loc { get; set; }
		public Member? Belong { get; set; }
		public Ident? Ident { get; set; }
		public IndentBlock? Body { get; set; }
	}
	public class Function : Member
	{
		public Loc Loc { get; set; }
		public Member? Belong { get; set; }
		public Ident? Ident { get; set; }
		public List<Field>? Parameters { get; set; }
		public Type? ReturnType { get; set; }
		public IndentBlock? Body { get; set; }
		public FunctionType? FuncType { get; set; }
		public bool IsCast { get; set; }
		public Loc CastLoc { get; set; }
	}
	public class IntType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public ulong BitSize { get; set; }
		public Endian Endian { get; set; }
		public bool IsSigned { get; set; }
	}
	public class IdentType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public Ident? Ident { get; set; }
		public Member? Base { get; set; }
	}
	public class IntLiteralType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public IntLiteral? Base { get; set; }
	}
	public class StrLiteralType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public StrLiteral? Base { get; set; }
	}
	public class VoidType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
	}
	public class BoolType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
	}
	public class ArrayType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public Loc EndLoc { get; set; }
		public Type? BaseType { get; set; }
		public Expr? Length { get; set; }
	}
	public class FunctionType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public Type? ReturnType { get; set; }
		public List<Type>? Parameters { get; set; }
	}
	public class StructType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public List<Member>? Fields { get; set; }
	}
	public class StructUnionType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public List<StructType>? Fields { get; set; }
		public Expr? Base { get; set; }
		public List<Field>? UnionFields { get; set; }
	}
	public class Cast : Expr
	{
		public Loc Loc { get; set; }
		public Type? ExprType { get; set; }
		public Call? Base { get; set; }
		public Expr? Expr { get; set; }
	}
	public class Comment : Node
	{
		public Loc Loc { get; set; }
		public string Comment_ { get; set; }
	}
	public class CommentGroup : Node
	{
		public Loc Loc { get; set; }
		public List<Comment>? Comments { get; set; }
	}
	public class UnionType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public Expr? Cond { get; set; }
		public List<UnionCandidate>? Candidates { get; set; }
		public StructUnionType? BaseType { get; set; }
	}
	public class UnionCandidate : Stmt
	{
		public Loc Loc { get; set; }
		public Expr? Cond { get; set; }
		public Member? Field { get; set; }
	}
	public class RangeType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public Type? BaseType { get; set; }
		public Range? Range { get; set; }
	}
	public class Enum : Member
	{
		public Loc Loc { get; set; }
		public Member? Belong { get; set; }
		public Ident? Ident { get; set; }
		public Scope? Scope { get; set; }
		public Loc ColonLoc { get; set; }
		public Type? BaseType { get; set; }
		public List<EnumMember>? Members { get; set; }
		public EnumType? EnumType { get; set; }
	}
	public class EnumMember : Member
	{
		public Loc Loc { get; set; }
		public Member? Belong { get; set; }
		public Ident? Ident { get; set; }
		public Expr? Expr { get; set; }
	}
	public class EnumType : Type
	{
		public Loc Loc { get; set; }
		public bool IsExplicit { get; set; }
		public Enum? Base { get; set; }
	}
	public enum UnaryOp
	{
		Not,
		MinusSign,
	}
	public enum BinaryOp
	{
		Mul,
		Div,
		Mod,
		LeftArithmeticShift,
		RightArithmeticShift,
		LeftLogicalShift,
		RightLogicalShift,
		BitAnd,
		Add,
		Sub,
		BitOr,
		BitXor,
		Equal,
		NotEqual,
		Less,
		LessOrEq,
		Grater,
		GraterOrEq,
		LogicalAnd,
		LogicalOr,
		CondOp1,
		CondOp2,
		RangeExclusive,
		RangeInclusive,
		Assign,
		DefineAssign,
		ConstAssign,
		AddAssign,
		SubAssign,
		MulAssign,
		DivAssign,
		ModAssign,
		LeftShiftAssign,
		RightShiftAssign,
		BitAndAssign,
		BitOrAssign,
		BitXorAssign,
		Comma,
	}
	public enum IdentUsage
	{
		Unknown,
		Reference,
		DefineVariable,
		DefineConst,
		DefineField,
		DefineFormat,
		DefineEnum,
		DefineEnumMember,
		DefineFn,
		DefineCastFn,
		DefineArg,
		ReferenceType,
	}
	public enum Endian
	{
		Unspec,
		Big,
		Little,
	}
	public enum TokenTag
	{
		Indent,
		Space,
		Line,
		Punct,
		IntLiteral,
		BoolLiteral,
		StrLiteral,
		Keyword,
		Ident,
		Comment,
		Error,
		Unknown,
	}
	public class Scope
	{
		public Scope? Prev { get; set; }
		public Scope? Next { get; set; }
		public Scope? Branch { get; set; }
		public List<Ident>? Ident { get; set; }
		public Node? Owner { get; set; }
	}
	public class Pos
	{
		public ulong Begin { get; set; }
		public ulong End { get; set; }
	}
	public class Loc
	{
		public Pos Pos { get; set; }
		public ulong File { get; set; }
		public ulong Line { get; set; }
		public ulong Col { get; set; }
	}
	public class Token
	{
		public TokenTag Tag { get; set; }
		public string Token_ { get; set; }
		public Loc Loc { get; set; }
	}
	public class SrcErrorEntry
	{
		public string Msg { get; set; }
		public string File { get; set; }
		public Loc Loc { get; set; }
		public string Src { get; set; }
		public bool Warn { get; set; }
	}
	public class SrcError
	{
		public List<SrcErrorEntry>? Errs { get; set; }
	}


}

File: ast2dart/lib/ast.dart
library ast;
enum NodeType {
Program,
Comment,
CommentGroup,
FieldArgument,
Expr,
Binary,
Unary,
Cond,
Ident,
Call,
If,
MemberAccess,
Paren,
Index,
Match,
Range,
TmpVar,
Import,
Cast,
Available,
SpecifyOrder,
ExplicitError,
IoOperation,
BadExpr,
Stmt,
Loop,
IndentBlock,
ScopedStatement,
MatchBranch,
UnionCandidate,
Return,
Break,
Continue,
Assert,
ImplicitYield,
Metadata,
Type,
IntType,
FloatType,
IdentType,
IntLiteralType,
StrLiteralType,
VoidType,
BoolType,
ArrayType,
FunctionType,
StructType,
StructUnionType,
UnionType,
RangeType,
EnumType,
MetaType,
OptionalType,
GenericType,
Literal,
IntLiteral,
BoolLiteral,
StrLiteral,
CharLiteral,
TypeLiteral,
SpecialLiteral,
Member,
Field,
Format,
State,
Enum,
EnumMember,
Function,
BuiltinMember,
BuiltinFunction,
BuiltinField,
BuiltinObject,
}
enum TokenTag {
Indent,
Space,
Line,
Punct,
IntLiteral,
BoolLiteral,
StrLiteral,
CharLiteral,
Keyword,
Ident,
Comment,
Error,
Unknown,
}
enum UnaryOp {
Not,
MinusSign,
}
enum BinaryOp {
Mul,
Div,
Mod,
LeftArithmeticShift,
RightArithmeticShift,
LeftLogicalShift,
RightLogicalShift,
BitAnd,
Add,
Sub,
BitOr,
BitXor,
Equal,
NotEqual,
Less,
LessOrEq,
Grater,
GraterOrEq,
LogicalAnd,
LogicalOr,
CondOp1,
CondOp2,
RangeExclusive,
RangeInclusive,
Assign,
DefineAssign,
ConstAssign,
AddAssign,
SubAssign,
MulAssign,
DivAssign,
ModAssign,
LeftShiftAssign,
RightShiftAssign,
BitAndAssign,
BitOrAssign,
BitXorAssign,
Comma,
}
enum IdentUsage {
Unknown,
Reference,
DefineVariable,
DefineConst,
DefineField,
DefineFormat,
DefineState,
DefineEnum,
DefineEnumMember,
DefineFn,
DefineCastFn,
DefineArg,
ReferenceType,
ReferenceMember,
ReferenceMemberType,
MaybeType,
ReferenceBuiltinFn,
}
enum Endian {
Unspec,
Big,
Little,
}
enum ConstantLevel {
Unknown,
Constant,
ImmutableVariable,
Variable,
}
enum BitAlignment {
ByteAligned,
Bit1,
Bit2,
Bit3,
Bit4,
Bit5,
Bit6,
Bit7,
NotTarget,
NotDecidable,
}
enum Follow {
Unknown,
End,
Fixed,
Constant,
Normal,
}
enum IoMethod {
Unspec,
OutputPut,
InputPeek,
InputGet,
InputBackward,
InputOffset,
InputBitOffset,
InputRemain,
InputSubrange,
ConfigEndianLittle,
ConfigEndianBig,
ConfigEndianNative,
ConfigBitOrderLsb,
ConfigBitOrderMsb,
}
enum SpecialLiteralKind {
Input,
Output,
Config,
}
enum OrderType {
Byte,
BitStream,
BitMapping,
BitBoth,
}
abstract class Node {
Loc loc = Loc();
}
abstract class Expr extends Node {
Type? exprType;
ConstantLevel constantLevel = ConstantLevel.Unknown;
}
abstract class Stmt extends Node {
}
abstract class Type extends Node {
bool isExplicit = false;
bool nonDynamicAllocation = false;
BitAlignment bitAlignment = BitAlignment.ByteAligned;
int? bitSize;
}
abstract class Literal extends Expr {
}
abstract class Member extends Stmt {
Member? belong;
StructType? belongStruct;
Ident? ident;
}
abstract class BuiltinMember extends Member {
}
class Program extends Node {
StructType? structType;
List<Node>? elements = [];
Scope? globalScope;
}
class Comment extends Node {
String comment = '';
}
class CommentGroup extends Node {
List<Comment>? comments = [];
}
class FieldArgument extends Node {
Expr? rawArguments;
Loc endLoc = Loc();
List<Expr>? collectedArguments = [];
List<Expr>? arguments = [];
Expr? alignment;
int? alignmentValue;
Expr? subByteLength;
Expr? subByteBegin;
Expr? peek;
int? peekValue;
TypeLiteral? typeMap;
List<Metadata>? metadata = [];
}
class Binary extends Expr {
BinaryOp op = BinaryOp.Mul;
Expr? left;
Expr? right;
}
class Unary extends Expr {
UnaryOp op = UnaryOp.Not;
Expr? expr;
}
class Cond extends Expr {
Expr? cond;
Expr? then;
Loc elsLoc = Loc();
Expr? els;
}
class Ident extends Expr {
String ident = '';
IdentUsage usage = IdentUsage.Unknown;
Node? base;
Scope? scope;
}
class Call extends Expr {
Expr? callee;
Expr? rawArguments;
List<Expr>? arguments = [];
Loc endLoc = Loc();
}
class If extends Expr {
StructUnionType? structUnionType;
Scope? condScope;
Expr? cond;
IndentBlock? then;
Node? els;
}
class MemberAccess extends Expr {
Expr? target;
Ident? member;
Ident? base;
}
class Paren extends Expr {
Expr? expr;
Loc endLoc = Loc();
}
class Index extends Expr {
Expr? expr;
Expr? index;
Loc endLoc = Loc();
}
class Match extends Expr {
StructUnionType? structUnionType;
Scope? condScope;
Expr? cond;
List<MatchBranch>? branch = [];
}
class Range extends Expr {
BinaryOp op = BinaryOp.Mul;
Expr? start;
Expr? end;
}
class TmpVar extends Expr {
int tmpVar = 0;
}
class Import extends Expr {
String path = '';
Call? base;
Program? importDesc;
}
class Cast extends Expr {
Call? base;
Expr? expr;
}
class Available extends Expr {
Call? base;
Expr? target;
}
class SpecifyOrder extends Expr {
Binary? base;
OrderType orderType = OrderType.Byte;
Expr? order;
int? orderValue;
}
class ExplicitError extends Expr {
Call? base;
StrLiteral? message;
}
class IoOperation extends Expr {
Expr? base;
IoMethod method = IoMethod.Unspec;
List<Expr>? arguments = [];
}
class BadExpr extends Expr {
String content = '';
}
class Loop extends Stmt {
Scope? condScope;
Expr? init;
Expr? cond;
Expr? step;
IndentBlock? body;
}
class IndentBlock extends Stmt {
StructType? structType;
List<Node>? elements = [];
Scope? scope;
}
class ScopedStatement extends Stmt {
StructType? structType;
Node? statement;
Scope? scope;
}
class MatchBranch extends Stmt {
Match? belong;
Expr? cond;
Loc symLoc = Loc();
Node? then;
}
class UnionCandidate extends Stmt {
Expr? cond;
Field? field;
}
class Return extends Stmt {
Expr? expr;
}
class Break extends Stmt {
}
class Continue extends Stmt {
}
class Assert extends Stmt {
Binary? cond;
bool isIoRelated = false;
}
class ImplicitYield extends Stmt {
Expr? expr;
}
class Metadata extends Stmt {
Expr? base;
String name = '';
List<Expr>? values = [];
}
class IntType extends Type {
Endian endian = Endian.Unspec;
bool isSigned = false;
bool isCommonSupported = false;
}
class FloatType extends Type {
Endian endian = Endian.Unspec;
bool isCommonSupported = false;
}
class IdentType extends Type {
MemberAccess? importRef;
Ident? ident;
Type? base;
}
class IntLiteralType extends Type {
IntLiteral? base;
}
class StrLiteralType extends Type {
StrLiteral? base;
StrLiteral? strongRef;
}
class VoidType extends Type {
}
class BoolType extends Type {
}
class ArrayType extends Type {
Loc endLoc = Loc();
Type? elementType;
Expr? length;
int? lengthValue;
}
class FunctionType extends Type {
Type? returnType;
List<Type>? parameters = [];
}
class StructType extends Type {
List<Member>? fields = [];
Node? base;
bool recursive = false;
int fixedHeaderSize = 0;
int fixedTailSize = 0;
}
class StructUnionType extends Type {
Expr? cond;
List<Expr>? conds = [];
List<StructType>? structs = [];
Expr? base;
List<Field>? unionFields = [];
bool exhaustive = false;
}
class UnionType extends Type {
Expr? cond;
List<UnionCandidate>? candidates = [];
StructUnionType? baseType;
Type? commonType;
}
class RangeType extends Type {
Type? baseType;
Range? range;
}
class EnumType extends Type {
Enum? base;
}
class MetaType extends Type {
}
class OptionalType extends Type {
Type? baseType;
}
class GenericType extends Type {
Member? belong;
}
class IntLiteral extends Literal {
String value = '';
}
class BoolLiteral extends Literal {
bool value = false;
}
class StrLiteral extends Literal {
String value = '';
int length = 0;
}
class CharLiteral extends Literal {
String value = '';
int code = 0;
}
class TypeLiteral extends Literal {
Type? typeLiteral;
Loc endLoc = Loc();
}
class SpecialLiteral extends Literal {
SpecialLiteralKind kind = SpecialLiteralKind.Input;
}
class Field extends Member {
Loc colonLoc = Loc();
bool isStateVariable = false;
Type? fieldType;
FieldArgument? arguments;
int? offsetBit;
int offsetRecent = 0;
int? tailOffsetBit;
int tailOffsetRecent = 0;
BitAlignment bitAlignment = BitAlignment.ByteAligned;
BitAlignment eventualBitAlignment = BitAlignment.ByteAligned;
Follow follow = Follow.Unknown;
Follow eventualFollow = Follow.Unknown;
Field? next;
}
class Format extends Member {
IndentBlock? body;
Func? encodeFn;
Func? decodeFn;
List<Func>? castFns = [];
List<IdentType>? depends = [];
List<Field>? stateVariables = [];
}
class State extends Member {
IndentBlock? body;
}
class Enum extends Member {
Scope? scope;
Loc colonLoc = Loc();
Type? baseType;
List<EnumMember>? members = [];
EnumType? enumType;
}
class EnumMember extends Member {
Expr? rawExpr;
Expr? value;
StrLiteral? strLiteral;
}
class Func extends Member {
List<Field>? parameters = [];
Type? returnType;
IndentBlock? body;
FunctionType? funcType;
bool isCast = false;
Loc castLoc = Loc();
}
class BuiltinFunction extends Member {
FunctionType? funcType;
}
class BuiltinField extends Member {
Type? fieldType;
}
class BuiltinObject extends Member {
List<BuiltinMember>? members = [];
}
class Scope {
Scope? prev;
Scope? next;
Scope? branch;
List<Ident>? ident = [];
Node? owner;
bool branchRoot = false;
}
class Pos {
int begin = 0;
int end = 0;
}
class Loc {
Pos pos = Pos();
int file = 0;
int line = 0;
int col = 0;
}
class Token {
TokenTag tag = TokenTag.Indent;
String token = '';
Loc loc = Loc();
}
class RawScope {
int? prev;
int? next;
int? branch;
List<int> ident = [];
int? owner;
bool branchRoot = false;
}
class RawNode {
NodeType nodeType = NodeType.Program;
Loc loc = Loc();
dynamic body;
}
class SrcErrorEntry {
String msg = '';
String file = '';
Loc loc = Loc();
String src = '';
bool warn = false;
}
class SrcError {
List<SrcErrorEntry> errs = [];
}
class JsonAst {
List<RawNode> node = [];
List<RawScope> scope = [];
}
class AstFile {
List<String> files = [];
JsonAst? ast;
SrcError? error;
}
class TokenFile {
List<String> files = [];
List<Token>? tokens = [];
SrcError? error;
}
